feat(dynamic sampling): Implements tasks for generating transaction boosting rules [TET-519]

RaduW opened the pull request:
This PR implements the infrastructure that gets the data for generating transaction boosting rules.
It implements the tasks that collect transaction boosting information and store it in redis

Commit:
asdasd

Commit:
asdasd

getsantry[bot] committed:
style(lint): Auto commit lint changes

Commit:
refactor a bit

Commit:
add more tests

Commit:
add more tests

Commit:
fixup!

Commit:
refactor test

Commit:
refactor test

Commit:
fix tests

Commit:
ignore test before snuba PR is ready

Commit:
add more tests

Commit:
add more tests

Commit:
adressed code review

Commit:
adressed code review

Commit:
add more tests

Commit:
add more tests

Commit:
add more tests

Commit:
add more tests

Commit:
fix mypy

Commit:
add more tests

Commit:
remove second utils.py

Commit:
rename to projects_with_tx_count

Commit:
fix MOCK_DATETIME

narsaynorath committed:
chore(view-hierarchy): Add proguard file size to span (#44826)

Tracking the file size so we can understand its impact on timing

Commit:
switch to redis HSET, rename test

Commit:
undo incorrect change

Commit:
add pexpire key

Commit:
refactor celery task

Commit:
add timeouts to celery tasks

Commit:
refactor using redis pipelines

Commit:
remove  "c:transactions/count_per_root_project@none" from relay list

RaduW committed:
WIP, implementation almost complete

RaduW committed:
fix typing

Commit:
add option to filter amount of orgs

Commit:
add more tests

Commit:
fix tests

Commit:
Merge branch 'master' into andrii/implement-cronjob-to-fetch-orgs-projects

Commit:
Merge branch 'master' into andrii/implement-cronjob-to-fetch-orgs-projects

Commit:
Merge branch 'master' into andrii/implement-cronjob-to-fetch-orgs-projects

Commit:
rewrite using indexer.resolve_shared_org()

Commit:
fix tests

Commit:
update referrer for prioritise by project bias

Commit:
Merge branch 'master' into andrii/implement-cronjob-to-fetch-orgs-projects

RaduW committed:
WIP removing direct dependencies on SHARED_STRINGS

RaduW committed:
minor, adjust code style to match prioritise by project volume

Commit:
Merge branch 'master' into feat/dynamic-sampling/prioritise-transactions-cronjob

getsantry[bot] committed:
style(lint): Auto commit lint changes

andriisoldatenko commented on src/sentry/dynamic_sampling/rules/helpers/prioritize_transactions.py:
```diff
@@ -0,0 +1,37 @@
+
+
+def _get_cache_key(org_id: int, proj_id: int) -> str:
+    return f"ds::o:{org_id}:p:{proj_id}:pri_tran"
```
remember we decided to use a short version (without project_id):
https://github.com/getsentry/sentry/blob/41ff6d0ca205a37ce74d49780781136467a5f7d6/src/sentry/dynamic_sampling/rules/helpers/prioritise_project.py#L9-L10

and later use `hset` instead.

RaduW committed:
Merge branch 'master' into feat/dynamic-sampling/prioritise-transactions-cronjob

RaduW committed:
WIP new fetching procedure for transaction rebalancing

RaduW committed:
WIP small cleanup

RaduW committed:
fix db related functions, add unit tests

RaduW committed:
convert ProjectTransactions to TypedDict, added task test

RaduW committed:
merge master

RaduW committed:
naming fixes

RaduW committed:
Merge branch 'master' into feat/dynamic-sampling/prioritise-transactions-cronjob

evanh commented on src/sentry/dynamic_sampling/tasks.py:
```diff
@@ -102,3 +117,90 @@ def adjust_sample_rates(
+                    current_org = Organization.objects.get_from_cache(
+                        id=project_transactions["org_id"]
+                    )
+                    current_org_enabled = features.has(
```
I would be inclined to move this check before the transactions queries. Why fetch and process all the transactions if we are just going to throw them away?

RaduW replied:
Thanks @evanh  for you input.
Yes, you are absolutely right, we have debated about checking early and getting the transactions only for the
orgs that are enabled.

The reason we decided to go for this is that we wanted to check if we are able to run the queries for everything and for now only apply the changes to a limited group of orgs (that we use for testing).

The feature is only intended to be temporary, once we are happy with the result on our test orgs we will enable this for everybody and we don't want to have the unpleasant surprise that the database cannot handle it.

If we have any doubts about the query performance we can use the same approach that we used for project and do the first query ( the query for orgs) using modulo i.e. `Condition(Function("modulo", [Column("org_id"), 100]), Op.LT, load_rate)`. This way we can ramp up db load independently from enabling the organisations for transaction boosting.

WDYT ?

RaduW replied:
Done.

evanh reviewed (commented):
This generally looks good to me. Added one small comment.

onewland commented on src/sentry/dynamic_sampling/prioritise_transactions.py:
```diff
@@ -0,0 +1,304 @@
+
+logger = logging.getLogger(__name__)
+MAX_SECONDS = 60
+CHUNK_SIZE = 1000
```
do we want to increase `CHUNK_SIZE` to 10000 (Snuba's limit)? or did we turn down that idea for some reason?

https://www.notion.so/sentry/Project-bias-analysis-44a0c93358c94979b18ebd06a6b7f52e?pvs=4#34a2cff938f04d34a4b62a1c75acaadd

RaduW replied:
Yes thanks, I missed that.

onewland reviewed (commented):
Mostly seems good to me, but if we're not increasing `CHUNK_SIZE` before merge I'd like to know the reason

RaduW committed:
added db ramp-up parameter for prioritise transaction

RaduW committed:
increased chunk size to 10000

nikhars commented on src/sentry/dynamic_sampling/prioritise_transactions.py:
```diff
@@ -0,0 +1,307 @@
+                    Column("org_id"),
+                ],
+                where=[
+                    Condition(Function("modulo", [Column("org_id"), 100]), Op.LT, load_rate),
```
This is fine for now. When you decide to change this to 100% and are satisfied with the results, please remove this modulo clause altogether.

RaduW replied:
Yes, thanks @nikhars, this is the plan.

nikhars reviewed (approved):
🔥

RaduW committed:
set CHUNK_SIZE to 9998 since Snuba's limit is 10000 and we need space

We fetch CHUNK_SIZE+1 so we need that to be <=10000
set it to 9998 to be sure we don't get at the limit.

andriisoldatenko commented on src/sentry/dynamic_sampling/prioritise_transactions.py:
```diff
@@ -0,0 +1,307 @@
+    current_proj_id: Optional[int] = None
+    transaction_counts: List[Tuple[str, int]] = []
+
+    if large_transactions:
```
```suggestion
    if large_transactions_first:
```

RaduW replied:
The fact that the large transactions come first is not the point, the point of the parameter is that the function fetches large_transactions ( i.e. the sorting is not relevant, what the function returns is).

andriisoldatenko commented on src/sentry/dynamic_sampling/rules/helpers/prioritize_transactions.py:
```diff
@@ -0,0 +1,37 @@
+            ret_val = cast(Tuple[Mapping[str, float], float], val)
+            return ret_val
+    except (TypeError, ValueError):
+        # for any error just get the default
```
do we what to capture exceptions here by sentry?

RaduW replied:
Don't remember exactly why I have both, but it only catches deserialisation errors.

andriisoldatenko replied:
`sentry_sdk.capture_exception(e)`

RaduW committed:
capture deserialisation errors in Sentry

RaduW committed:
Merge branch 'master' into feat/dynamic-sampling/prioritise-transactions-cronjob
